tracing/ops_tracing/_backend.py

```diff
@@ -0,0 +1,132 @@
+
+if TYPE_CHECKING:
+    from ops.jujucontext import _JujuContext
+
+from ._const import BUFFER_FILE, Config
+from ._export import BufferingSpanExporter
+
+
+class LogsToEvents(logging.Handler):
+    """An adaptor that convert log records to OTEL events."""
+
+    def emit(self, record: logging.LogRecord) -> None:
+        """Emit this log record as OTEL event."""
+        span = get_current_span()
+        if span and span.is_recording():
+            try:
+                message = record.getMessage()
+                level = record.levelname
+            except Exception as e:
+                # This should never happen, except if the charm includes a custom logging
+                # library like structlog that enriches both the format and record attributes,
+                # or if the record format doesn't match the provided arguments.
+                message = f'log {record=} error {e}'
+                level = 'UNKNOWN'
+            span.add_event(message, {'level': level})
+
+
+@contextlib.contextmanager
+def setup(juju_context: _JujuContext, charm_class_name: str) -> Generator[None, None, None]:
+    """Control tracing lifespan of tracing.
+
+    Args:
+        juju_context: the context for this dispatch, for annotation
+        charm_class_name: the name of the charm class, for annotation
+    """
+    app_name, unit_number = juju_context.unit_name.split('/', 1)
+    # Note that the Resource is immutable, and we want to start tracing early.
+    # This means that the Charmhub charm name (self.meta.name) is not available yet.
+    resource = Resource.create(
+        attributes={
+            'service.namespace': juju_context.model_uuid,
+            'service.namespace.name': juju_context.model_name,
+            'service.name': app_name,
+            'service.instance.id': unit_number,
+            'service.charm': charm_class_name,
+        }
+    )
+    set_tracer_provider(_create_provider(resource, juju_context.charm_dir))
+    logging.root.addHandler(log_handler := LogsToEvents())
+    try:
+        yield
+    finally:
+        logging.root.handlers.remove(log_handler)
+        shutdown_tracing()
+
+
+def _create_provider(resource: Resource, charm_dir: pathlib.Path) -> TracerProvider:
+    """Create the OpenTelemetry tracer provider."""
+    # Separate function so that it's easy to override in tests
+    exporter = BufferingSpanExporter(charm_dir / BUFFER_FILE)
+    span_processor = BatchSpanProcessor(exporter)
+    return TracerProvider(resource=resource, active_span_processor=span_processor)  # type: ignore
+
+
+def get_exporter() -> BufferingSpanExporter | None:
+    """Get our export from OpenTelemetry SDK."""
+    try:
+        exporter = get_tracer_provider()._active_span_processor.span_exporter  # type: ignore
+    except AttributeError:
+        # The global tracer provider was not configured by us and has a wrong processor.
+        return
```

Comment:
Nit: it's returning None rather than just returning nothing, so let's be explicit (here and below):
```suggestion
        return None
```